keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,148 @@
+---
+kep-number: 0008
+title: Sidecar Containers
+authors:
+  - "@joseph-irving"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+  - sig-node
+reviewers:
+  - "@fejta"
+approvers:
+  - "@enisoc"
+editor: TBD
+creation-date: 2018-05-14
+last-updated: 2018-11-05
+status: provisional
+---
+
+# Sidecar Containers
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives)
+
+## Summary
+
+To solve the problem of container lifecycle dependency we can create a new class of container: a "sidecar container" that behaves primarily like a normal container but is handled differently during termination and startup.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, they are becoming more common in a lot of applications and as more people have used them, more issues have cropped up.
+
+Here are some examples of the main problems:
+
+### Jobs
+ If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
```

Comment:
> your sidecar container will carry on running so the job will never finish

That's a separate problem and specific to jobs. 
1. 'Sidecar' in most cases should be started before and terminated after the primaryContainer (ordering)
2. A successful exit of a container should hook the termination of other containers. 

Do we want to combine both logic, 'ordering', 'hook termination'

Reply:
The comment was more about having two separated fields to express both problem:
e.g:
```
 - name: app1
   lifecycleBound: 
      - name: app2
      - name: app3
 - name: app2
 - name: app3
```
in this case, if app2 AND app3 receive a signal or has already terminated, the sidecar(app1) exits too.
There's no notion of order, just trigger termination. 

in parallel of this notation, we could have fields to express the order.